projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -287,46 +333,29 @@ internal static InboundFrame ReadFrom(Stream reader, byte[] frameHeaderBuffer, u
             const int EndMarkerLength = 1;
             // Is returned by InboundFrame.ReturnPayload in Connection.MainLoopIteration
             int readSize = payloadSize + EndMarkerLength;
-            byte[] payloadBytes = ArrayPool<byte>.Shared.Rent(readSize);
-            try
-            {
-                ReadFromStream(reader, payloadBytes, readSize);
-            }
-            catch (Exception)
-            {
-                // Early EOF.
-                ArrayPool<byte>.Shared.Return(payloadBytes);
-                throw new MalformedFrameException($"Short frame - expected to read {readSize} bytes");
-            }
 
-            if (payloadBytes[payloadSize] != Constants.FrameEnd)
+            if ((buffer.Length - 7) < readSize)
             {
-                ArrayPool<byte>.Shared.Return(payloadBytes);
-                throw new MalformedFrameException($"Bad frame end marker: {payloadBytes[payloadSize]}");
+                frame = default;
+                return false;
             }
-
-            RabbitMqClientEventSource.Log.DataReceived(payloadSize + Framing.BaseFrameSize);
-            return new InboundFrame(type, channel, new Memory<byte>(payloadBytes, 0, payloadSize), payloadBytes);
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static void ReadFromStream(Stream reader, byte[] buffer, int toRead)
-        {
-            int bytesRead = 0;
-            do
+            else
             {
-                int read = reader.Read(buffer, bytesRead, toRead - bytesRead);
-                if (read == 0)
+                byte[] payloadBytes = ArrayPool<byte>.Shared.Rent(readSize);
+                ReadOnlySequence<byte> framePayload = buffer.Slice(7, readSize);
+                framePayload.CopyTo(payloadBytes);
+
+                if (payloadBytes[payloadSize] != Constants.FrameEnd)
                 {
-                    ThrowEndOfStream();
+                    ArrayPool<byte>.Shared.Return(payloadBytes);
+                    throw new MalformedFrameException($"Bad frame end marker: {payloadBytes[payloadSize]}");
```

Comment:
I know we return without clearing but couldn't we theoretically read garbage when throwing in cases when the returned buffer is already used again by another thread that rented it? Hw about reading the end marker first and then returning?

Reply:
It seems reasonable to clear the array in this case.

Reply:
My comment was not about clearing the array but about the order.

We return `payloadBytes` to the pool which makes it available for use and then throw but access in the exception constructor the end marker with `payloadBytes[payloadSize]` accessing the buffer we have just returned which might already be used by some other thread.

I was asking if the code should look like

```csharp
var endMarker = payloadBytes[payloadSize];
ArrayPool<byte>.Shared.Return(payloadBytes);
throw new MalformedFrameException($"Bad frame end marker: {endMarker}");
```

Reply:
OH! I see 👍